sofar/sofar.py

```diff
@@ -0,0 +1,2125 @@
+            f"(SOFA version {self.GLOBAL_Version})\n")
+        info_str += "-" * len(info_str) + "\n"
+
+        if info in ["all", "mandatory", "optional", "read only", "data"]:
+
+            info_str += f"showing {info} entries : type (shape), flags\n\n"
+
+            for key in keys:
+
+                # check if field should be skipped
+                flags = self._convention[key]["flags"]
+                if (not self._mandatory(flags) and info == "mandatory") \
+                        or \
+                        (self._mandatory(flags) and info == "optional") \
+                        or \
+                        (not self._read_only(flags) and info == "read only") \
+                        or \
+                        (self._convention[key]['type'] == "attribute" and
+                         info == "data"):
+                    continue
+
+                info_str += f"{key} : {self._convention[key]['type']}"
+
+                if self._convention[key]['dimensions']:
+                    info_str += \
+                        f" ({self._convention[key]['dimensions'].upper()})"
+
+                if self._mandatory(flags):
+                    info_str += ", mandatory"
+                else:
+                    info_str += ", optional"
+                if self._read_only(flags):
+                    info_str += ", read only"
+
+                if self._convention[key]['comment']:
+                    info_str += f"\n    {self._convention[key]['comment']}\n"
+                else:
+                    info_str += "\n"
+
+        elif info in keys:
+
+            for key in [k for k in keys if info in k]:
+                comment = str(self._convention[key]['comment'])
+
+                info_str += (
+                    f"{key}\n"
+                    f"    type: {self._convention[key]['type']}\n"
+                    f"    mandatory: "
+                    f"{self._mandatory(self._convention[key]['flags'])}\n"
+                    f"    read only: "
+                    f"{self._read_only(self._convention[key]['flags'])}\n"
+                    f"    default: {self._convention[key]['default']}\n"
+                    f"    shape: "
+                    f"{str(self._convention[key]['dimensions']).upper()}\n"
+                    f"    comment: {comment}\n")
+        else:
+            raise ValueError(f"info='{info}' is invalid")
+
+        print(info_str)
+
+    def add_variable(self, name, value, dtype, dimensions):
+        """
+        Add custom variable to the SOFA object, i.e., numeric or string arrays.
+
+        Parameters
+        ----------
+        name : str
+            Name of the new variable.
+        value : any
+            value to be added (see `dtype` for restrictions).
+        dtype : str
+            Type of the entry to be added in netCDF style:
+
+            ``'double'``
+                Use this to store numeric data that can be provided as number
+                list or numpy array.
```

Comment:
Does netCDF always use `double` precision by default, or could it also be possible to use single precision? (or general float precisions, aka python builtins, or np datatypes)

Reply:
netCDF could do all this, SOFA specifies all numeric variables to be double (probably due to the Matlab origin of the API)